### PR TITLE
dns: log queries as an array - v3

### PR DESCRIPTION
**Pull request by jasonish**
Previous PR: https://github.com/OISF/suricata/pull/11238

Changes from previous PR:
- unify the code for logging DNS request across DNS and alerts, addresses discrepancies reported in bug
- documentation/upgrade guide

Ticket: https://redmine.openinfosecfoundation.org/issues/6281

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1907

### CODE REVIEW

File: rust/src/dns/log.rs [hunk 1]
```diff
@@ -584,92 +550,132 @@ fn dns_log_json_answer(
...
+        return Ok(());
+    };
+
+    // The internal Suricata transaction ID.
+    jb.set_uint("tx_id", tx.id - 1)?;
+
+    // The on the wire DNS transaction ID.
+    jb.set_uint("id", tx.tx_id() as u64)?;
+
+    // Log header fields. Should this be a sub-object?
```
Comment (jasonish) [inline rust/src/dns/log.rs:593]: Oops, forgot to do this: https://github.com/OISF/suricata/pull/11238#discussion_r1626682580

**Comment by jasonish** [comment]
Rendered upgrade guide (@jufajardini): https://jasonish-suricata.readthedocs.io/en/dns-request-array-v3/upgrade/8.0-dns-logging-changes.html

File: doc/userguide/upgrade/8.0-dns-logging-changes.rst [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+                 "flags": "100",
+                 "rd": true,
+                 "opcode": 0,
+                 "rcode": "NOERROR",
+                 "queries": [
+                   {
+                     "rrname": "www.suricata.io",
+                     "rrtype": "A"
+                   }
+                 ]
```
Comment (jasonish) [inline doc/userguide/upgrade/8.0-dns-logging-changes.rst:62]: This format is critical for mDNS as well.

Comment (jasonish) [reply doc/userguide/upgrade/8.0-dns-logging-changes.rst:62]: Elasticsearch is pretty with arrays like this, you can search on `dns.queries.rrname`. So it means searches/reports, etc will need to change from `dns.rrname` to `dns.queries.rrname`.

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 21035

File: rust/src/dns/log.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/dns/log.rs:593]: So, is this PR a draft ?

File: doc/userguide/upgrade/8.0-dns-logging-changes.rst [hunk 3]
```diff
@@ -0,0 +1,260 @@
+:orphan: Document not referenced in a toctree, so add this.
```
Comment (catenacyber) [inline doc/userguide/upgrade/8.0-dns-logging-changes.rst:1]: What does this `orphan` mean ?

File: rust/src/dns/log.rs [hunk 4]
```diff
@@ -584,92 +550,132 @@ fn dns_log_json_answer(
...
+        jb.close()?;
     }
 
-    return Ok(false);
-}
+    // For now, stop logging if a request to limit changes from 7.0
+    // behavior. But revisit as it could be useful to log this data
+    // when its part of a request.
+    if tx.is_request() {
+        return Ok(());
```
Comment (catenacyber) [inline rust/src/dns/log.rs:635]: This would be just extending...
I would log this data if any...

File: rust/src/dns/log.rs [hunk 5]
```diff
@@ -584,92 +550,132 @@ fn dns_log_json_answer(
...
+        jb.open_array("authorities")?;
+        for auth in &message.authorities {
+            let auth_detail = dns_log_json_answer_detail(auth)?;
+            jb.append_object(&auth_detail)?;
         }
+        jb.close()?;
     }
+
+    // TODO:
+    // - additionals
```
Comment (catenacyber) [inline rust/src/dns/log.rs:652]: Is there a ticket for this ?

File: etc/schema.json [hunk 6]
```diff
@@ -1101,7 +1101,7 @@
                 "authorities": {
                     "$ref": "#/$defs/dns.authorities"
                 },
-                "query": {
+                "queries": {
                     "type": "array",
```
Comment (catenacyber) [inline etc/schema.json:1105]: Did you run a check if the schema is complete ?
cf 9332bc2c45f1e1e08f2261857d5804f41ce7e639

**Comment by catenacyber** [comment]
Looks a nice clean up, thanks for taking on this

File: doc/userguide/upgrade/8.0-dns-logging-changes.rst (hunk 3 above)
Comment (jasonish) [reply doc/userguide/upgrade/8.0-dns-logging-changes.rst:1]: Sphinx will complain if a document exists but is not reference in a toctree (table of contents). This document is only linked from its upgrade note, so we suppress the error with this.

File: rust/src/dns/log.rs (hunk 4 above)
Comment (jasonish) [reply rust/src/dns/log.rs:635]: Yes, another reason back to draft. I do want to log that but it involved even more test changes, so this helped reduce the changes.

File: rust/src/dns/log.rs (hunk 5 above)
Comment (jasonish) [reply rust/src/dns/log.rs:652]: Yes, and a PR: https://github.com/OISF/suricata/pull/11263

File: etc/schema.json (hunk 6 above)
Comment (jasonish) [reply etc/schema.json:1105]: It passes existing tests. Do we have another check to run?

File: doc/userguide/upgrade/8.0-dns-logging-changes.rst [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+       - 8.0
+     * - .. code-block::
+
+           {
+             "event_type": "dns",
+             "dns": {
+               "version": 2,
+               "type": "answer",
+               "id": 0,
+               "flags": "8180",
```
Comment (jasonish) [inline doc/userguide/upgrade/8.0-dns-logging-changes.rst:83]: Any thoughts on if flags would look better as:
```
"flags": ["qr", "rd", "ra"],
```

File: etc/schema.json (hunk 6 above)
Comment (catenacyber) [reply etc/schema.json:1105]: No tests, but as you reread the code, maybe you found more that are missing and not existing in tests...

File: rust/src/dns/log.rs (hunk 4 above)
Comment (catenacyber) [reply rust/src/dns/log.rs:635]: How come this involves test changes ? This is just extension of the logs, right ?...

File: doc/userguide/upgrade/8.0-dns-logging-changes.rst (hunk 7 above)
Comment (catenacyber) [reply doc/userguide/upgrade/8.0-dns-logging-changes.rst:83]: Looks better as ` ["qr", "rd", "ra"],`  than `8180` as a string ?

File: etc/schema.json (hunk 6 above)
Comment (catenacyber) [reply etc/schema.json:1105]: Looks like `SOA` is missing from code review

File: doc/userguide/upgrade/8.0-dns-logging-changes.rst (hunk 7 above)
Comment (jasonish) [reply doc/userguide/upgrade/8.0-dns-logging-changes.rst:83]: We'd probably want to keep the raw value (`8180` in this case).  The array would replace `"rd": true` and friends.

File: rust/src/dns/log.rs (hunk 4 above)
Comment (jasonish) [reply rust/src/dns/log.rs:635]: > How come this involves test changes ? This is just extension of the logs, right ?...

Perhaps not this, but when I had a ton of failures, it was simpler to reduce the scope to no new behavior.

File: etc/schema.json (hunk 6 above)
Comment (jasonish) [reply etc/schema.json:1105]: > Looks like `SOA` is missing from code review

We don't have any tests that log it outside of authorities.  And I know there's been some question whether we should have it in the schema if we don't have a test that exercises it.  Thoughts?

File: doc/userguide/upgrade/8.0-dns-logging-changes.rst (hunk 7 above)
Comment (catenacyber) [reply doc/userguide/upgrade/8.0-dns-logging-changes.rst:83]: ok as well

File: etc/schema.json (hunk 6 above)
Comment (catenacyber) [reply etc/schema.json:1105]: Last time, I ended up creating a command for sshfp by running a dig command...

Comment (jasonish) [reply etc/schema.json:1105]: Ok, I ended up with some when I added logging all arrays for requests.  A zone update request will log them.

File: rust/src/dns/log.rs (hunk 5 above)
Comment (jasonish) [reply rust/src/dns/log.rs:652]: New PR has this work merged in.

File: rust/src/dns/log.rs (hunk 4 above)
Comment (jasonish) [reply rust/src/dns/log.rs:635]: Will be in next version.

File: doc/userguide/upgrade/8.0-dns-logging-changes.rst (hunk 7 above)
Comment (victorjulien) [reply doc/userguide/upgrade/8.0-dns-logging-changes.rst:83]: How would that affect things like elastic and splunk wrt querying for them?

Comment (jasonish) [reply doc/userguide/upgrade/8.0-dns-logging-changes.rst:83]: Currently you'd have to do something like:
```dns.flags.rd:true dns.flags.z:true```
with this change:
```dns.flags:rd dns.flags:z```

Comment (victorjulien) [reply doc/userguide/upgrade/8.0-dns-logging-changes.rst:83]: Makes sense, thanks

**Pull request closed without merging**